voltdbclient/client.go

```diff
@@ -0,0 +1,240 @@
+type Client struct {
+	config       *ClientConfig
+	reader       io.Reader
+	writer       io.Writer
+	connData     *connectionData
+	netListener  *NetworkListener
+	clientHandle int64
+}
+
+// connectionData are the values returned by a successful login.
+type connectionData struct {
+	hostId      int32
+	connId      int64
+	leaderAddr  int32
+	buildString string
+}
+
+type ClientConfig struct {
+	username string
+	password string
+}
+
+type NullValue struct {
+	colType int8
+}
+
+type Callback struct {
+	Channel <-chan *Response
+	Handle  int64
+}
+
+// NewClient creates an initialized, authenticated Client.
+func NewClient(username string, password string) *Client {
+	var client = new(Client)
+	client.config = &ClientConfig{username, password}
+	return client
+}
+
+func NewCallback(channel <-chan *Response, handle int64) *Callback {
+	var callback = new(Callback)
+	callback.Channel = channel
+	callback.Handle = handle
+	return callback
+}
+
+func NewNullValue(colType int8) *NullValue {
+	var nv = new(NullValue)
+	nv.colType = colType
+	return nv
+}
+
+// Call invokes the procedure 'procedure' with parameter values 'params'
+// and returns a pointer to the received Response.
+func (client *Client) Call(procedure string, params ...interface{}) (*Response, error) {
+	if client.writer == nil {
+		return nil, fmt.Errorf("Can not call procedure on closed Client.")
+	}
+	handle := atomic.AddInt64(&client.clientHandle, 1)
+	cb := client.netListener.registerCallback(handle)
+	if err := client.writeProcedureCall(client.writer, procedure, handle, params); err != nil {
+		client.netListener.removeCallback(handle)
+		return nil, err
+	}
+	return <-cb.Channel, nil
+}
+
+// CallAsync asynchronously invokes the procedure 'procedure' with parameter values 'params'.
+// A pointer to the Response from the server will be put on the returned channel.
+func (client *Client) CallAsync(procedure string, params ...interface{}) (*Callback, error) {
+	if client.writer == nil {
+		return nil, fmt.Errorf("Can not call procedure on closed Client.")
+	}
+	handle := atomic.AddInt64(&client.clientHandle, 1)
+	cb := client.netListener.registerCallback(handle)
+	if err := client.writeProcedureCall(client.writer, procedure, handle, params); err != nil {
+		client.netListener.removeCallback(handle)
+		return nil, err
+	}
+	return cb, nil
+}
+
+func (client *Client) CreateConnection(hostAndPort string) error {
+	raddr, err := net.ResolveTCPAddr("tcp", hostAndPort)
+	if err != nil {
+		return fmt.Errorf("Error resolving %v.", hostAndPort)
+	}
+	var tcpConn *net.TCPConn
+	if tcpConn, err = net.DialTCP("tcp", nil, raddr); err != nil {
+		return err
+	}
+	client.reader = tcpConn
+	client.writer = tcpConn
+	login, err := serializeLoginMessage(client.config.username, client.config.password)
+	if err != nil {
+		return err
+	}
+	client.writeLoginMessage(&login)
+	if client.connData, err = client.readLoginResponse(); err != nil {
```

Comment:
if there is error or login failed server will close. Do we have to do any cleanup on client side for tcpConn Or client itself?
How does user know that this connection is gone.